### ISSUE TITLE
[Clang] Unsafe pointers and C standard issues

### ISSUE THREAD

**Issue by luc0x61** [llvm/llvm-project#63326]
The following short code exploits **three** possible issues with warnings and errors:

```c
/// \file  unsafe.c
// clang -std=c11 -Weverything unsafe.c -o unsafe.exe
#include  <stddef.h>

static int  goo (int * pi, size_t a)
{
    return pi[a];
}

int  main (int argc, char *argv[])
{
    (void)argc;
    (void)argv;
    int     goz[4] = { 0, };
    
    return goo(goz, 0);
}
```

Compiling on Windows 11 pro x64, but with the standard x86 installer (released Clang/LLVM 16.0.6, `LLVM-16.0.6-win32.exe`):
```console
# clang --version
clang version 16.0.6
Target: i686-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files (x86)\LLVM\bin

# clang -std=c11 -Weverything unsafe.c -o unsafe.exe
unsafe.c:5:24: warning: 'pi' is an unsafe pointer used for buffer access
      [-Wunsafe-buffer-usage]
static int  goo (int * pi, size_t a)
                 ~~~~~~^~
unsafe.c:7:12: note: used in buffer access here
    return pi[a];
           ^~
unsafe.c:14:13: warning: mixing declarations and code is incompatible with
      standards before C99 [-Wdeclaration-after-statement]
    int     goz[4] = { 0, };
            ^
2 warnings generated.
```

1. Looks like I can't pass a generic pointer and do any arithmetics on it without the unsafe pointer warning (lines 5, 7): am I missing anything?
2. I requested C11 standard, but still receive C99 warning (line 14): does it make any sense?
3. Note that if I delete/comment the `#include` line there is no error/warning on `size_t` missing definition, compilation goes smooth

**Comment by AaronBallman** [comment]
> Looks like I can't pass a generic pointer and do any arithmetics on it without the unsafe pointer warning (lines 5, 7): am I missing anything?

Correct, that's specified by the RFC: https://discourse.llvm.org/t/rfc-c-buffer-hardening/65734 I brought up concerns about exposing this in C (CC @jkorous-apple @haoNoQ for awareness), but the feeling was that this still has value in C code bases. You've opted in to this warning by passing `-Weverything`.

> I requested C11 standard, but still receive C99 warning (line 14): does it make any sense?

Yes, you opted into compatibility warnings by passing `-Weverything` and those include compatibility with previous standards. We're inconsistent about previous standard compatibility warnings currently, it's a newer diagnostic push.

> Note that if I delete/comment the #include line there is no error/warning on size_t missing definition, compilation goes smooth

Because you're compiling on Windows, clang automatically enables `-fms-compatibility` which is why you're seeing the behavior you're seeing: https://godbolt.org/z/reGd1xsoY

**Comment by AaronBallman** [comment]
> @AaronBallman thank you for your answers and the details!

Any time!

> I can understand the first point, about pointers arithmetic, but IMHO there can't be applied safe access staying with C language. There are limits with it, pushing beyond which we just turn it into _another language_, breaking compatibility.

The buffer safety work is conforming to the best of my knowledge -- it does not change the meaning of valid C code, it only recommends changes. The [bounds safety RFC](https://discourse.llvm.org/t/rfc-enforcing-bounds-safety-in-c-fbounds-safety/70854), as a counter-example, does change the meaning of valid C code and that's why it is behind a feature flag that users must opt into explicitly. When a feature is creating another language mode, we typically put it behind a feature flag so that `-std=<lang mode> -pedantic-errors` can be the most strictly-conforming mode we can give users in terms of extension use (and `-Wall -Wextra` help pick up the non-extension issues that have a fairly low false-positive rate).

> I had some suspicions about the second point on standards, and have been addressed with some questions that I raised on the StackOverflow discussion platform.
> 
> I will then avoid the `-Weverything` as suggested in the manual, and use `-Wall -Wextra`, unless when I want to have a deeper _opinion_ from the compiler.

Personally, I recommend never using `-Weverything` (I think it was a mistake for us to expose it as a warning flag). It is a lie -- it doesn't enable *all* warnings (for example, passing `-fsyntax-only -Weverything` will miss codegen and LLVM backend warnings), the warnings it does enable are sometimes contradictory, and it enables at least one warning that changes the meaning of code in C (`-Wwrite-strings`).

**Comment by spaesani** [comment]
Interestingly some people have no issue performing in depth static analysis
For over 20 years now
https://en.m.wikipedia.org/wiki/Astr%C3%A9e_(static_analysis)

Maybe clang needs a fork and some commercialisation to bring it up to date.

**Comment by tstellar** [comment]
@spaesani Please review our [Code of Conduct](https://llvm.org/docs/CodeOfConduct.html)

**Comment by spaesani** [comment]
And you are?Sent from my Galaxy
-------- Original message --------From: Tom Stellard ***@***.***> Date: 2024-09-04  1:20 PM  (GMT-05:00) To: llvm/llvm-project ***@***.***> Cc: spaesani ***@***.***>, Mention ***@***.***> Subject: Re: [llvm/llvm-project] [Clang] Unsafe pointers and C standard issues (Issue #63326) 
@spaesani Please review our Code of Conduct

**Comment by spaesani** [comment]
@tstellar Start by identifying yourself in relation to github and say what you allege violates "our" , implying your, code of conduct.

**Comment by tstellar** [comment]
@spaesani This is the bug tracker for the [LLVM](https://llvm.org) project.  All users are required to follow the LLVM project's [Code of Conduct](https://llvm.org/docs/CodeOfConduct.html).  Since you are new to the community, I recommend that you familiarize yourself with the policy and ensure that all your communication going forward adheres to our rules.  Specifically, some of your previous comments were not respectful or kind.

I have no special authority when it comes to the Code of Conduct, I am just trying to be helpful so you can participate in our community if you want to.

**Comment by spaesani** [comment]
This is github.
I agreed to their terms of service not yours.

Now, if you want to add to the conversation and answer how you expect clang developers to magically  detect an out of bounds access with an attribute they specify while not being able to do so with a C specified method of specifying an array size ie with static initialization by all means go ahead because as the op has made apparent for all to see that's as much of a bug as you can possibly have in your compiler and as you said this is an issue discussion. On github not llvm.org.
Try respecting the users of their platform.
Starting by respecting. theiir terms of use. Not trying to impose yours

**Comment by jkorous-apple** [comment]
@spaesani Constructive feedback or better yet patches with improvements are always welcome! :)
For discussion of the approach in general a much better venue would be https://discourse.llvm.org/ ; this is off-topic here.

**Comment by spaesani** [comment]
.Your dingenousness is glaring. Much like the issues with your compiler.

The issue is open here as are the examples provided by op and this iz where they posted the issue. It is obviously the place to discuss it and answer the question I posed that you're now evading.
Far from excellant.
https://docs.github.com/en/site-policy/github-terms/github-terms-of-service

**Comment by AaronBallman** [comment]
Closing this issue as the questions appear to have all been answered. If there are more questions in this space, feel free to reopen with a comment.
